pkg/filter/expr_filter.go

```diff
@@ -0,0 +1,374 @@
+	case row.IsUpdate():
+		oldExprs, err := r.getUpdateOldExpr(ti)
+		if err != nil {
+			return false, err
+		}
+		newExprs, err := r.getUpdateNewExpr(ti)
+		if err != nil {
+			return false, err
+		}
+		ignoreOld, err := r.skipDMLByExpression(row.RowChangedDatums.PreRowDatums, oldExprs)
+		if err != nil {
+			return false, err
+		}
+		ignoreNew, err := r.skipDMLByExpression(row.RowChangedDatums.RowDatums, newExprs)
+		if err != nil {
+			return false, err
+		}
+		return ignoreOld || ignoreNew, nil
+	case row.IsDelete():
+		exprs, err := r.getDeleteExpr(ti)
+		if err != nil {
+			return false, err
+		}
+		return r.skipDMLByExpression(row.RowChangedDatums.PreRowDatums, exprs)
+	default:
+		log.Warn("unknown row changed event type")
+		return false, nil
+	}
+}
+
+func (r *dmlExprFilterRule) skipDMLByExpression(rowData []types.Datum,
+	expr expression.Expression,
+) (bool, error) {
+	if len(rowData) == 0 || expr == nil {
+		return false, nil
+	}
+
+	row := chunk.MutRowFromDatums(rowData).ToRow()
+
+	d, err := expr.Eval(row)
+	if err != nil {
+		log.Error("failed to eval expression", zap.Error(err))
+		return false, errors.Trace(err)
+	}
+	if d.GetInt64() == 1 {
+		return true, nil
+	}
+	return false, nil
+}
+
+func getColumnFromError(err error) string {
+	if !core.ErrUnknownColumn.Equal(err) {
+		return err.Error()
+	}
+	column := strings.TrimSpace(strings.TrimPrefix(err.Error(),
+		"[planner:1054]Unknown column '"))
+	column = strings.TrimSuffix(column, "' in 'expression'")
+	return column
+}
+
+// dmlExprFilter is a filter that filters DML events by SQL expression.
+type dmlExprFilter struct {
+	rules []*dmlExprFilterRule
+}
+
+func newExprFilter(timezone string, cfg *config.FilterConfig) (*dmlExprFilter, error) {
+	res := &dmlExprFilter{}
+	sessCtx := utils.NewSessionCtx(map[string]string{
+		"time_zone": timezone,
+	})
+	for _, rule := range cfg.EventFilters {
+		err := res.addRule(sessCtx, rule)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return res, nil
+}
+
+func (f *dmlExprFilter) addRule(sessCtx sessionctx.Context,
+	cfg *config.EventFilterRule,
+) error {
+	rule, err := newExprFilterRule(sessCtx, cfg)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	f.rules = append(f.rules, rule)
+	return nil
+}
+
+// verify checks if all rules in this filter is valid.
+func (f *dmlExprFilter) verify(tableInfos []*model.TableInfo) error {
+	for _, rule := range f.rules {
+		err := rule.verify(tableInfos)
+		if err != nil {
+			log.Error("failed to verify expression filter rule", zap.Error(err))
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+func (f *dmlExprFilter) getRules(schema, table string) []*dmlExprFilterRule {
+	res := make([]*dmlExprFilterRule, 0)
+	for _, rule := range f.rules {
+		if rule.tableMatcher.MatchTable(schema, table) {
+			res = append(res, rule)
+		}
+	}
+	return res
+}
+
+// shouldSkipDML skips dml event by sql expression.
+func (f *dmlExprFilter) shouldSkipDML(row *model.RowChangedEvent,
```

Comment:
Could you add a benchmark? Since it applies to all changes, it may become a bottleneck.

Reply:
yes.

Reply:
// cmd: go test -benchmem -run=^$ -bench ^BenchmarkSkipDML$ github.com/pingcap/tiflow/pkg/filter
// goos: maxOS 12.3.1
// goarch: arm64
// cpu: Apple M1 Pro
// BenchmarkSkipDML/insert-1
// BenchmarkSkipDML/insert-1-10              990166              1151 ns/op             768 B/op         24 allocs/op
// BenchmarkSkipDML/insert-2
// BenchmarkSkipDML/insert-2-10             1000000              1187 ns/op             768 B/op         24 allocs/op
// BenchmarkSkipDML/update
// BenchmarkSkipDML/update-10                698208              1637 ns/op            1480 B/op         43 allocs/op
// BenchmarkSkipDML/delete
// BenchmarkSkipDML/delete-10               1000000              1112 ns/op             768 B/op         24 allocs/op